server/src/test/java/org/elasticsearch/index/mapper/DocumentParserTests.java

```diff
@@ -707,12 +715,15 @@ public void testDeepSubfieldAfterSubfieldMappingUpdate() throws Exception {
     public void testObjectMappingUpdate() throws Exception {
         MapperService mapperService = createMapperService();
         DocumentMapper docMapper = mapperService.documentMapper();
-        List<Mapper> updates = new ArrayList<>();
-        updates.add(createObjectMapper("foo"));
-        updates.add(createObjectMapper("foo.bar"));
-        updates.add(new MockFieldMapper("foo.bar.baz"));
-        updates.add(new MockFieldMapper("foo.field"));
-        Mapping mapping = DocumentParser.createDynamicUpdate(docMapper.mappers(), updates, Collections.emptyList());
+        ParsedDocument doc = docMapper.parse(source(b -> {
+            b.startObject("foo");
+            b.startObject("bar");
+            b.field("baz", 10);
+            b.endObject();
+            b.field("field", 10);
+            b.endObject();
+        }));
```

Comment:
In your opinion, does changing these tests from directly testing DocumentParser.createDynamicUpdate() to parsing full json objects also change their coverage? As far as I understand, we previously sorted all mapper updates from a parse alphabetically (thus the sorting in the test setups). Now we don't need to do that anymore due to the new recursive strategy. However, I wonder if my re-ordering things in the json we might create a different update order and maybe different results/edge cases? I'm not sure this can happen so just wondering if that might be a concern.

Reply:
The main difference I think is if we have two incompatible dynamic mappers for the same field - let's say you have something that looks like this:
```
{ "foo" : { "bar" : 1 }, "foo.bar" : "baz" }
```
We'll generate two mappings for 'foo.bar' here, one `long` and one `text`+`keyword`.  Previously they would have been merged together and thrown an exception; now one will overwrite the other, but we'll still get an exception when the document is re-parsed if the `long` update is seen second and therefore 'wins'.  I think this is OK though?

This also opens up possibilities for automatic widening of dynamic types - we could in the future have some kind of merge detection that says if you have a `long` mapper and a `text` mapper, you should always accept the `text` mapper when you combine the two in a dynamic context.